### ISSUE TITLE
Refine_facets_3::number_of_bad_elements_impl() is not up-to-date (won't compile)

### ISSUE THREAD

**Issue by goran-w** [CGAL/cgal#3553]
## Issue Details

In CGAL 4.13 I'm trying to call `Mesher_3<C3T3,MC,MD>::display_number_of_bad_elements()` which won't compile, due to out-dated code in `Refine_facets_3<Tr,Cr,MD,C3T3_,P_,Ct,B_,C_>::number_of_bad_elements_impl()`.

I'm using the following kernel etc:

```
using K = CGAL::Exact_predicates_inexact_constructions_kernel;
using MD = CGAL::Mesh_domain_with_polyline_features_3<CGAL::Labeled_mesh_domain_3<K>>;
using Tr = CGAL::Mesh_triangulation_3<MD, K, CGAL::Parallel_tag>::type;
using C3T3 = CGAL::Mesh_complex_3_in_triangulation_3<Tr>;
using MC = CGAL::Mesh_criteria_3<Tr>;
```

In `Refine_facets_3::number_of_bad_elements_impl()` the following issues need to be fixed:

* The calls to ` r_criteria_(...)` are missing their first argument which should be `r_tr_`.

* The variables storing the result from` r_oracle_.is_in_domain_object()(...)` are declared as `bool` but should be changed to `MD::Subdomain` (which usually is `boost::optional<int>`), and the subsequent printing could be fixed by including `<boost/optional/optional_io.hpp>`.

## Environment

* Operating system: Windows 10 64 bits
* Compiler: Microsoft Visual Studio Professional 2017 version 15.9.4
* Release or debug mode: (Any)
* Specific flags used (if any):
* CGAL version: 4.13
* Boost version: 1.67.0
* Other libraries versions if used (Eigen, TBB, etc.): Eigen 3.3.5, tbb2019_20181010oss_win

**Comment by lrineau** [comment]
Thanks @goran-w. If you have a version of the code that works, would you mind create a pull-request?

**Comment by goran-w** [comment]
Hi! I created pull request #3556. I'm not sure how you'd like to handle the MD::Subdomain return values, maybe you can take a look. I made the minimal changes needed to make it compile.

To provoke the compiler errors, simply call display_number_of_bad_elements() on a Mesher_3 object.

**Comment by lrineau** [comment]
Thanks for the PR @goran-w.

I did not even know about this function `display_number_of_bad_elements()`. It was probably a debug function, added by one of the collaborators of CGAL Mesh_3.

I had a look at the PR and approved it. It will be tested in our nightly testsuite, and eventually merged.

**Comment by goran-w** [comment]
Thanks! The crux is that my fix will not be tested unless this function is being called somewhere, since the compiler will not visit it due to templates not being instantiated.

I discovered this bug while experimenting with the Step-by-step methods in the Mesher_3 class:

```
  // Step-by-step methods
  void initialize();
  void fix_c3t3();
  void display_number_of_bad_elements();
  void one_step();
  bool is_algorithm_done();
  struct Mesher_status {...}
```

I'm trying to go a little bit more low-level with the tet-mesher (similar to CGAL/demo/Polyhedron/Plugins/Mesh_3/Mesh_function.h).

**Comment by goran-w** [comment]
BTW, I discovered that after using `Mesher_3::one_step()` and `Mesher_3::fix_c3t3()` I had to call `C3T3::rescan_after_load_of_triangulation()`, because otherwise CGAL::output_to_medit() could produce an inconsistent output file (mismatch between stated number of tris/tets and the actual number of tri/tet rows). 

This in turn was due to mismatch between the cached number of facets/cells and the actual enumerated (is_in_complex filtered) facets/cells in C3T3.

**Comment by goran-w** [comment]
Should it be considered a bug (new issue) that `CGAL::output_to_medit()` is "capable" of producing an incorrect .mesh file (due to the above mentioned potential cached-number mismatch)? 

For example, I got one file which reported 86 tetrahedra but there were no lines for those:

```
...
Tetrahedra
86
End
```

In another case, the .mesh file reported something like 560 tetrahedra but there were only like 263 lines.

**Comment by lrineau** [comment]
Fixed by #3553. The test still has to be written.
